src/__tests__/useAsync-test.js

```diff
@@ -0,0 +1,156 @@
+/* eslint react/prop-types: 0 */
+import '@testing-library/jest-dom/extend-expect';
+import * as React from 'react';
+import '@babel/polyfill';
+import { render, waitForDomChange, fireEvent, act } from '@testing-library/react';
+import useAsync from '../useAsync';
+
+function DisplayState({ api }) {
+  const result = {
+    value: api.value(),
+    success: api.success(),
+    error: api.error(),
+    initial: api.initial(),
+    loading: api.loading(),
+
+    hasValue: api.hasValue(),
+    isSuccess: api.isSuccess(),
+    isLoading: api.isLoading(),
+    isError: api.isError(),
+    isInitial: api.isInitial(),
+
+    matches: api.match({
+      default: () => ({ type: 'default' }),
+      loading: () => ({ type: 'loading' }),
+      success: value => ({ type: 'success', value }),
+      error: msg => ({ type: 'error', value: msg.toString() }),
```

Comment:
`error: value => ({ type: 'error', value })`, combined with the other requested change.